presto-main/src/main/java/com/facebook/presto/sql/planner/iterative/trait/CardinalityTrait.java

```diff
@@ -0,0 +1,123 @@
+/*
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+package com.facebook.presto.sql.planner.iterative.trait;
+
+import com.facebook.presto.sql.planner.iterative.Trait;
+import com.facebook.presto.sql.planner.iterative.TraitType;
+import com.google.common.collect.Range;
+
+import java.util.Objects;
+import java.util.OptionalLong;
+
+import static com.google.common.base.MoreObjects.toStringHelper;
+import static java.util.Objects.requireNonNull;
+
+public class CardinalityTrait
+        implements Trait<CardinalityTrait>
+{
+    public static CardinalityTrait scalar()
+    {
+        return exactly(1);
+    }
+
+    public static CardinalityTrait exactly(long cardinality)
+    {
+        return new CardinalityTrait(Range.singleton(cardinality));
+    }
+
+    public static CardinalityTrait atMostScalar()
+    {
+        return atMost(1L);
+    }
+
+    public static CardinalityTrait atMost(long cardinality)
+    {
+        return new CardinalityTrait(Range.closed(0L, cardinality));
+    }
+
+    private final Range<Long> cardinalityRange;
+
+    public CardinalityTrait(Range<Long> cardinalityRange)
```

Comment:
Is accepting anything other than closed-end ranges useful?
This looks elegant, but I'm afraid this only introduces unnecessary complexity and hides the fact we instantiate only closed-ranges.